rust/ql/test/library-tests/type-inference/main.rs

```diff
@@ -1927,6 +1948,10 @@ mod impl_trait {
         let a = get_a_my_trait(); // $ target=get_a_my_trait
         let c = uses_my_trait2(a); // $ type=c:S2 target=uses_my_trait2
         let d = uses_my_trait2(S1); // $ type=d:S2 target=uses_my_trait2
+        let e = get_a_my_trait2(S1).get_a(); // $ target=get_a_my_trait2 target=MyTrait::get_a type=e:S1
+        // For this function the `impl` type does not appear in the root of the return type
+        let f = get_a_my_trait3(S1).unwrap().get_a(); // $ target=get_a_my_trait3 target=unwrap target=MyTrait::get_a type=f:S1
+        let g = get_a_my_trait4(S1).0.get_a(); // $ target=get_a_my_trait4 target=MyTrait::get_a type=g:S1
```

Comment:
I'm interested in what happens when we have nested `impl` types, for example I've seen:
```
impl Iterator<Item = (impl Into<String>, Resource)>
```
I haven't tested anything but it feels like there's a danger the type parameter `String` could become associated with both `impl` types on this line due to the way `implTraitTypeParam` is written?  Is it even wrong if it does???

Reply:
This is a good question that I haven't thought of.

With how it works right now they would become type parameters of both the `impl` types. Whether that's right or wrong I'm not sure about. My guess is that it's more correct to only create a type parameter of the inner `impl`.  I can try and write some tests for that in a follow up PR and take a look?

Reply:
In a follow-up PR is fine, if you're keeping track. :)

Reply:
I've added an internal issue for this.